hugr-passes/src/dataflow/datalog.rs

```diff
@@ -0,0 +1,346 @@
+            if let Some(tailloop) = ctx.get_optype(*tl).as_tail_loop(),
+            input_child(tl, in_n),
+            output_child(tl, out_n),
+            node_in_value_row(out_n, out_in_row), // get the whole input row for the output node...
+            // ...and select just what's possible for CONTINUE_TAG, if anything
+            if let Some(fields) = out_in_row.unpack_first(TailLoop::CONTINUE_TAG, tailloop.just_inputs.len()),
+            for (out_p, v) in fields.enumerate();
+
+        // Output node of child region propagate to outputs of tail loop
+        out_wire_value(tl, OutgoingPort::from(out_p), v) <-- node(tl),
+            if let Some(tailloop) = ctx.get_optype(*tl).as_tail_loop(),
+            output_child(tl, out_n),
+            node_in_value_row(out_n, out_in_row), // get the whole input row for the output node...
+            // ... and select just what's possible for BREAK_TAG, if anything
+            if let Some(fields) = out_in_row.unpack_first(TailLoop::BREAK_TAG, tailloop.just_outputs.len()),
+            for (out_p, v) in fields.enumerate();
+
+        // Conditional --------------------
+        // <Node> is a `Conditional` and its <usize>'th child (a `Case`) is <Node>:
+        relation case_node(Node, usize, Node);
+        case_node(cond, i, case) <-- node(cond),
+          if ctx.get_optype(*cond).is_conditional(),
+          for (i, case) in ctx.children(*cond).enumerate(),
+          if ctx.get_optype(case).is_case();
+
+        // inputs of conditional propagate into case nodes
+        out_wire_value(i_node, OutgoingPort::from(out_p), v) <--
+          case_node(cond, case_index, case),
+          input_child(case, i_node),
+          node_in_value_row(cond, in_row),
+          let conditional = ctx.get_optype(*cond).as_conditional().unwrap(),
+          if let Some(fields) = in_row.unpack_first(*case_index, conditional.sum_rows[*case_index].len()),
+          for (out_p, v) in fields.enumerate();
+
+        // outputs of case nodes propagate to outputs of conditional *if* case reachable
+        out_wire_value(cond, OutgoingPort::from(o_p.index()), v) <--
+          case_node(cond, _i, case),
+          case_reachable(cond, case),
+          output_child(case, o),
+          in_wire_value(o, o_p, v);
+
+        // In `Conditional` <Node>, child `Case` <Node> is reachable given our knowledge of predicate:
+        relation case_reachable(Node, Node);
+        case_reachable(cond, case) <-- case_node(cond, i, case),
+            in_wire_value(cond, IncomingPort::from(0), v),
+            if v.supports_tag(*i);
+
+        // CFG --------------------
+        relation cfg_node(Node); // <Node> is a `CFG`
+        cfg_node(n) <-- node(n), if ctx.get_optype(*n).is_cfg();
+
+        // In `CFG` <Node>, basic block <Node> is reachable given our knowledge of predicates:
+        relation bb_reachable(Node, Node);
+        bb_reachable(cfg, entry) <-- cfg_node(cfg), if let Some(entry) = ctx.children(*cfg).next();
```

Comment:
Yes, fair. I think the same applies to conditional (if any non-predicate input is bottom, no cases are reachable; if the elements of any variant contain bottom, that case is unreachable), Call, DataflowBlock/ExitBlock (in the same way as conditional, i.e. gate every CF edge), and depending on semantics perhaps also DFG (i.e. if a nested DFG is a scheduling barrier)

Reply:
Done by augmenting `ValueRow::unpack_first` (now `ValueRow::unpack_first_no_bottom`) and cases for CFG+DFG

Reply:
Ah, crossed-in-post with your lattice suggestion.
Yes, I've been thinking for awhile that a proper hierarchical notion of reachability would be good.
For starters, this could generalize `bb_reachable` and `case_reachable`.
Your idea of gating `in_wire_value` on reachability is quite cool (so if one input is bottom, they all are), although might make flattening conditionals more difficult....and we can't make `in_wire_value` depend on `node_in_value_row` in circular fashion like that, so we'd need

`in_wire_value_pre` and `node_in_value_row_pre` *ignoring* reachability,
PLUS `node_in_value_row` that copies `node_in_value_row` if none of the elements are bottom (and the parent/case/bb are reachable)....I guess `in_wire_value` could read from `node_in_value_row(n)[port]`.

For this PR or later, hmmm.